overlord/state/change.go

```diff
@@ -282,32 +282,105 @@ func init() {
 	}
 }
 
+func (c *Change) isTaskWaiting(visited map[string]bool, t *Task) bool {
+	if ok := visited[t.ID()]; ok {
+		// if we run into cyclic things, then no
+		return false
+	}
+	visited[t.ID()] = true
+
+	for _, wt := range t.WaitTasks() {
+		switch wt.Status() {
+		case WaitStatus:
+			return true
+		case DoStatus:
+			// If it has waiters that are in 'Do', we must check whether
+			// that task is in turn waiting for something else.
+			if c.isTaskWaiting(visited, wt) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// isTaskReadyToUndo returns whether a task in undo is ready to run.
+func (c *Change) isTaskReadyToUndo(t *Task) bool {
+	for _, wt := range t.HaltTasks() {
+		if !wt.Status().Ready() {
+			return false
+		}
+	}
+	return true
+}
+
+// isChangeWaiting returns whether a change has all of its tasks blocked by
+// wait tasks. If just one of the tasks is able to run, then this returns
+// false.
+func (c *Change) isChangeWaiting() bool {
+	visited := make(map[string]bool)
+	for _, t := range c.Tasks() {
+		// All tasks in 'Do' must be able to be traced back
+		// to a task in 'Wait'. Otherwise the change is able
+		// to run.
+		switch t.Status() {
+		case DoStatus:
+			if !c.isTaskWaiting(visited, t) {
+				return false
+			}
+		case UndoStatus:
+			// If something is in undo status, check if it can
+			// run.
+			if c.isTaskReadyToUndo(t) {
+				return false
+			}
```

Comment:
there's an asymmetry between Undo and Do, why can't the logic be for Do be as simple, or otoh why the logic for Undo doesn't need to be as complicated?

Reply:
Super good point, after you mentioned this I revisited my logic here, and it only took moving a few things around to break it. I had oversimplified my approach to the Undo logic.